### ISSUE TITLE
basename: true seems to break non-basename patterns

### ISSUE THREAD

**Issue by arcanis** [micromatch/micromatch#162]
## Please describe the **minimum necessary steps** to reproduce this issue:

```
var micromatch = require("micromatch")
console.log(micromatch.isMatch('/lib/libzip.js', ['/lib/*'], {basename: true}));
console.log(micromatch.isMatch('/lib/libzip.js', ['/lib/*']));
```

## What is happening that shouldn't be?

```
false
true
```

The `basename` option seems to prevent `/lib/libzip.js` from matching against `/lib/*.js`. It seems to have been working-ish in the `3.1.10` (it prints `true / undefined` there), but not since the `4.0.0`.

## What should be happening instead?

```
true
true
```

**Comment by arcanis** [comment]
Workaround:

```js
function isMatch(cwd: string, patterns: Array<string>) {
  let paths = patterns;
  let basenames = [];

  for (let t = 0; t < patterns.length; ++t) {
    if (patterns[t].includes(`/`)) {
      if (paths !== patterns) {
        paths.push(patterns[t]);
      }
    } else {
      if (paths === patterns)
        paths = patterns.slice(0, t);

      basenames.push(patterns[t]);
    }
  }

  if (mm.isMatch(cwd, paths as any))
    return true;

  if (mm.isMatch(cwd, basenames as any, {basename: true}))
    return true;

  return false;
}
```

**Comment by jonschlinkert** [comment]
That seems correct to me. You're trying to match `libzip.js` with `/lib/*`. 

There are probably better ways to do what you're trying to do. Can you provide more detail about the use case?

**Comment by arcanis** [comment]
I'm not sure I understand - per the documentation:

> patterns without slashes will be matched against the basename of the path if it contains slashes

So when matching `/lib/libzip.js` against `/lib/*`, since the pattern contains slashes, then I don't expect the behavior to change compared to `basename` not being set.

In my case I'm applying a set of user-provided patterns to filter a list of files. Those patterns may or may not contain basename checks (think of a `.gitignore` format).

**Comment by jonschlinkert** [comment]
> I'm not sure I understand - per the documentation:

Ah, the documentation is worded poorly, but it's still correct. It should be reworded to simply: "patterns will be matched against the basename of the path".

> In my case I'm applying a set of user-provided patterns to filter a list of files. Those patterns may or may not contain basename checks (think of a `.gitignore` format).

It seems like you should be matching against the relative or absolute path, not basename. For example, you'll also need to take into account globs with leading slashes, which indicates an absolute pattern (incidentally, the opposite of what you want to accomplish with `basename`).

**Comment by arcanis** [comment]
Let's say we have the following gitignore:

```
/lib/foo.js
*.ts
```

In this case, `/lib/foo.js` must be compared against the absolute paths (ie `/src/lib/foo.js` shouldn't match) while `*.ts` should be matched against the basename (ie `/src/foo.ts` should match). From my understanding of the doc, both of those should be possible with `basename: true`.

If it's not the case, how do you suggest I get the right results for those two patterns without doing a function similar to the one I posted?

**Comment by jonschlinkert** [comment]
Sorry for the late reply @arcanis

## tldr;

> how do you suggest I get the right results for those two patterns without doing a function similar to the one I posted?

Since your example is limited to a single filepath and pattern, show with and without the `basename` option, I might be missing something, but it seems like you should get what you want by prefixing the pattern with `**`.

If I'm missing something, it might help to provide more examples and/or just point me to the code where you're using micromatch and I'd be happy to dive in further. 

## More detail

> Let's say we have the following gitignore:

To be clear, gitignore uses different rules than bash glob, so in your example, both `/lib/foo.js` and `*.ts` would match paths relative to the directory of the `.gitignore` file. 

So while `*.ts` in a gitignore file would match `/Users/dev/someuser/foo/bar.ts`, to achieve the same thing with a glob you would need to do `**/*.ts`.

> In this case, `/lib/foo.js` must be compared against the absolute paths (ie `/src/lib/foo.js` shouldn't match) 

`/lib/foo.js` wouldn't match whether `basename` was true or not. But `**/lib/foo.js` would match.

> while `*.ts` should be matched against the basename (ie `/src/foo.ts` should match)

Only if `basename` is true. 

When `basename` is true, file paths are converted to basenames. Thus, `/lib/*`, for example, would not match the file path `/lib/foo.js` since the path is first converted to `foo.js`.

```js
const { isMatch } = require("micromatch");

console.log(isMatch('/lib/libzip.js', ['/lib/*'], {basename: true})); // should be false
console.log(isMatch('/lib/libzip.js', ['/lib/*'])); // should be true
```

**Comment by rally25rs** [comment]
@jonschlinkert 

> Ah, the documentation is worded poorly, but it's still correct. It should be reworded to simply: "patterns will be matched against the basename of the path, regardless of whether or not the pattern contains a slash".

I ran into this same confusion almost a year later, and the docs are still the same (incorrect) wording.

**Comment by jonschlinkert** [comment]
@rally25rs pr?

**Comment by rally25rs** [comment]
@jonschlinkert #201

**Comment by Blasz** [comment]
> Ah, the documentation is worded poorly, but it's still correct.

I disagree, the documentation is incorrect for a couple of reasons:

1. The option only mentions globs without slashes, therefore it's reasonable to assume that the default behaviour occurs when globs do have slashes.
2. The docs for the option also state:
> Same behavior as minimatch option matchBase.

which is false. Minimatch's option behaves as per the desired behaviour in the OP where the documentation makes sense.

Imo it doesn't make sense for the option to not work at all for glob patterns with slashes.

**Comment by jonschlinkert** [comment]
> Imo it doesn't make sense for the option to not work at all for glob patterns with slashes.

You're saying that when I use `**/foo/bar/*.js` it should match basenames when `basename` is true? If that's what you're saying I disagree. 

Closing since I think this option is behaving as expected.
